### PR TITLE
cni: cleanup exclusion logic

### PR DESCRIPTION
**Pull request by howardjohn**
* Fix https://github.com/istio/istio/issues/44938
* Split out fix for https://github.com/istio/istio/issues/44537
* General refactoring to avoid deep nesting and improve logging

**Please provide a description of this PR:**

### CODE REVIEW

File: cni/pkg/plugin/kubernetes.go [hunk 1]
```diff
@@ -67,47 +69,43 @@ func getK8sPodInfo(client *kubernetes.Clientset, podName, podNamespace string) (
...
-		pi.InitContainers[initContainer.Name] = struct{}{}
-	}
-	for containerIdx, container := range pod.Spec.Containers {
-		log.Debugf("Inspecting pod %v/%v container %v", podNamespace, podName, container.Name)
-		pi.Containers[containerIdx] = container.Name
-
-		if container.Name == "istio-proxy" {
+	for _, c := range containers(pod) {
+		pi.Containers.Insert(c.Name)
+		if c.Name == "istio-proxy" {
```
Comment (bleggett) [inline cni/pkg/plugin/kubernetes.go:79]:
```suggestion
		if c.Name == ISTIO_NODEPROXY {
```

meganit: `const` as elsewhere in the package? maybe global to `plugin` package?

File: cni/pkg/plugin/kubernetes.go [hunk 2]
```diff
@@ -67,47 +69,43 @@ func getK8sPodInfo(client *kubernetes.Clientset, podName, podNamespace string) (
...
 
 	return pi, nil
 }
 
+func containers(pod *v1.Pod) []v1.Container {
+	res := make([]v1.Container, 0, len(pod.Spec.Containers)+len(pod.Spec.InitContainers))
+	res = append(res, pod.Spec.InitContainers...)
+	res = append(res, pod.Spec.Containers...)
+	return res
+}
```
Comment (bleggett) [inline cni/pkg/plugin/kubernetes.go:104]: I get that these are now combined for back compat with both KEP-style and non-KEP-style sidecar detection, but can we comment here stating that _eventually_ we want the KEP approach of init containers to be the sidecar default, and ultimately we should only be looking at init containers?

File: cni/pkg/plugin/plugin.go [hunk 3]
```diff
@@ -33,19 +33,22 @@ import (
...
 
-	interceptRuleMgrType   = defInterceptRuleMgrType
 	podRetrievalMaxRetries = 30
 	podRetrievalInterval   = 1 * time.Second
 )
 
-const ISTIOINIT = "istio-init"
+const (
+	ISTIOINIT  = "istio-init"
+	ISTIOPROXY = "istio-proxy"
```
Comment (bleggett) [inline cni/pkg/plugin/plugin.go:50]: Se prev. - should use the constant everywhere.

**Review by bleggett — APPROVED** [review]
Minor comments

File: cni/pkg/plugin/kubernetes.go (hunk 2 above)
Comment (howardjohn) [reply cni/pkg/plugin/kubernetes.go:104]: even if it's the default we will support the old one. I won't say forever but with 4 old k8s versions supported, optimistically we are looking at 7 k8s versions of 2.5years min before we can remove the old mode.
I'll make the comments more clear

**Review by bleggett — APPROVED** [review]

**Pull request merged**